examples/helloworld-tutorial.md

```diff
@@ -0,0 +1,379 @@
+syntax = "proto3";
+package helloworld;
+
+service Greeter {
+    rpc SayHello (HelloRequest) returns (HelloReply);
+}
+
+message HelloRequest {
+   string name = 1;
+}
+
+message HelloReply {
+    string message = 1;
+}
+```
+
+## Application Setup 
+Now that we have defined the protobuf for our application we can start writing our OCaml. First we need to add our required dependencies to the `dune-project`.
+
+``` shell
+(lang dune 3.7)
+
+(name helloworld_grpc)
+(generate_opam_files true)
+
+(package
+ (name helloworld_grpc)
+ (synopsis "Helloworld in gRPC")
+ (description "Helloworld in gRPC.")
+ (depends
+  (grpc-lwt (>= 0.1.0))
+  (h2-lwt-unix (>= 0.9.0))
+  (ocaml-protoc (>= 2.4)))
+```
+
+We include `ocaml-protoc-plugin` as a useful way to incorporate the generation of our client and server gRPC code into the build process of our application. We will setup this build process now:
+
+## Generating Server and Client code
+
+In the proto directory of your project create a `dune` file and add the following code:
+
+``` shell
+(library
+ (name helloworld)
+ (libraries ocaml-protoc))
+
+(rule
+ (targets
+   helloworld_pb.ml
+   helloworld_pb.mli
+   helloworld_pp.ml
+   helloworld_pp.mli
+   helloworld_types.ml
+   helloworld_types.mli)
+ (action
+  (run ocaml-protoc -binary -pp -ml_out ./ helloworld.proto))
+ (deps helloworld.proto))
+```
+
+Make sure the dependencies are installed
+``` shell
+$ opam install dune grpc-lwt.0.1.0 ocaml-protoc.2.4 -y
+```
+
+This build configuration file tells dune to compile your protobugs when you build your OCaml project. It uses the `ocaml-protoc` cli tool and bundles up the OCaml code as a library called `helloworld`, which you depend on when writing the gRPC server and client.
+
+## Writing our Server
+
+Now that the build process is written and our dependencies are all setup, we can begin writing the fun stuff! We need to import the things we will be using in our server, including the protobuf. Start by making a file called server.ml in your `/lib` directory and writing the following code:
+
+``` ocaml
+open Grpc_lwt
+```
+
+Next up, we implement the Greeter service we previously defined in our `.proto` file. He is what that might look like:
+
+``` ocaml
+let say_hello buffer =
+  let decoder = Pbrt.Decoder.of_string buffer in
+  let req = Helloworld.Helloworld_pb.decode_hello_request decoder in
+  let message =
+    if req.name = "" then "You forgot your name!"
+    else Format.sprintf "Hello, %s!" req.name
+  in
+  let reply = Helloworld.Helloworld_types.default_hello_reply ~message () in
+  let encoder = Pbrt.Encoder.create () in
+  Helloworld.Helloworld_pb.encode_hello_reply reply encoder;
+  Lwt.return (Grpc.Status.(v OK), Some (Pbrt.Encoder.to_string encoder))
+
+let greeter_service =
+  Server.Service.(
+    v () |> add_rpc ~name:"SayHello" ~rpc:(Unary say_hello) |> handle_request)
+
+let server =
+  Server.(
+    v () |> add_service ~name:"helloworld.Greeter" ~service:greeter_service)
```

Comment:
Relating to my above comment, with `ocaml-protoc-plugin` one can actually get the RPC and service names from the `.proto` file which is quite neat. (Mentioning this as a sidenote, not necessarily for inclusion in the tutorial -- we were considering writing a somewhat opinionated wrapper that would remove a lot of boilerplate)

Reply:
> we were considering writing a somewhat opinionated wrapper that would remove a lot of boilerplate.

That would be a good improvement for the API in my opinion, however I've only recently started using this library and don't have a considered opinion on how that should look.

Reply:
The service name and function name are squashed into the `val name` field on the generated code.  eg
``` ocaml
    module GetFeature = struct
      let name = "/routeguide.RouteGuide/GetFeature"
      module Request = Point
      module Response = Feature
    end      
```
What about making a change to the `Message` module type to include returning the components?

``` ocaml
module type Message = sig
  type t
  val name : () -> string  (* returning GetFeature*)
  val service_name: () -> string (* returning routeguide.RouteGuide *)
  val from_proto: Reader.t -> t Result.t
  val to_proto: t -> Writer.t
end
```

Reply:
> What about making a change to the `Message` module type to include returning the components?

This is a good idea, and it's not hard to do. In fact, we already did something like that in our fork. I could have sworn it was part of the changes that have been merged recently, but alas it wasn't. Here's the commit for reference:
https://github.com/dialohq/ocaml-protoc-plugin/commit/beeffb2fb3e522550918c1d54f033dfc2f0482be

There have been a few changes upstream so it won't merge cleanly anymore. I'll try to rebase it and open a new PR.

Reply:
That would be great, let me know if I can help with it.

Reply:
https://github.com/issuu/ocaml-protoc-plugin/pull/49